src/libsystemd/sd-journal/journal-file.c

```diff
@@ -3654,6 +3655,41 @@ int journal_file_open(
         return r;
 }
 
+int journal_file_parse_uid(const char *path, uid_t *ret_uid) {
+        _cleanup_free_ char *buf = NULL, *p = NULL;
+        const char *a, *b, *at;
+        int r;
+
+        /* This helper returns -EREMOTE when the filename doesn't match user online/offline journal
+         * pattern. Hence it currently doesn't parse archived or disposed user journals. */
+
+        assert(path);
+        assert(ret_uid);
+
+        r = path_extract_filename(path, &p);
+        if (r < 0)
+                return r;
+        if (r == O_DIRECTORY)
+                return -EISDIR;
+
+        a = startswith(p, "user-");
+        if (!a)
+                return -EREMOTE;
+        b = endswith(p, ".journal");
+        if (!b)
+                return -EREMOTE;
+
+        at = strchr(a, '@');
+        if (at)
+                return -EREMOTE;
+
+        buf = strndup(a, b-a);
+        if (!buf)
+                return -ENOMEM;
+
+        return parse_uid(buf, ret_uid);
```

Comment:
in case of parse failures here we probably should also return EREMOTE

Reply:
I'm not sure as we wouldn't be able to differentiate between the case where a file name is not matching the user file name pattern and the case where the pased uid is invalid. In the former case we should simply skip the file while for the latter case we should warn.